sql: fix COPY when it accesses TxnTimestamp

yuzefovich opened the pull request:
This commit fixes a bug introduced in 888001367dee4da574386488590b9234ff0481c5 where a call to `initPlanner` was added. In particular, the problem is as follows: in the beginning of `insertRowsInternal` we set up the planner in `preparePlannerForCopy` (which correctly sets the TxnTimestamp field of the eval context), and that commit added a call to `initPlanner` which deeply-updates the eval context, without setting all of the fields (TxnTimestamp is made nil). As a result, later on if we try to evaluate one of the builtin functions that need the TxnTimestamp, then we hit a nil pointer. For example, this can happen when we have a DEFAULT expression like `now()` for one of the columns. This is now fixed by calling `initPlanner` right away, when setting up the copy machine. Additionally, this commit adds this call when creating the copy file upload machine too.

Fixes: #104456.

Release note (bug fix): CockroachDB previously could encounter `zero transaction timestamp in EvalContext` when evaluating COPY command which copies to the table that has DEFAULT expression using time-related builtins like `now()` and `localtimestamp`. The bug was introduced in 23.1.0 and is now fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105233)
<!-- Reviewable:end -->

yuzefovich commented:
@rafiss @cucaroach could you help me out to come up with a regression test? We need a COPY statement that uses a builtin like `now()` or `localtimestamp()` and goes through "CopyIn" path. I looked at existing tests (in `sql/copy/copy_test.go` and `pgtest/copy`) and couldn't figure out how to adjust them for what we want.

yuzefovich commented:
Nevermind, I found a repro.

yuzefovich committed:
sql: fix COPY when it accesses TxnTimestamp

This commit fixes a bug introduced in 888001367dee4da574386488590b9234ff0481c5
where a call to `initPlanner` was added. In particular, the problem is
as follows: in the beginning of `insertRowsInternal` we set up the
planner in `preparePlannerForCopy` (which correctly sets the
TxnTimestamp field of the eval context), and that commit added a call to
`initPlanner` which deeply-updates the eval context, without setting all
of the fields (TxnTimestamp is made nil). As a result, later on if we
try to evaluate one of the builtin functions that need the TxnTimestamp,
then we hit a nil pointer. For example, this can happen when we have
a DEFAULT expression like `now()` for one of the columns. This is now
fixed by calling `initPlanner` right away, when setting up the copy
machine. Additionally, this commit adds this call when creating the copy
file upload machine too.

Release note (bug fix): CockroachDB previously could encounter
`zero transaction timestamp in EvalContext` when evaluating COPY command
which copies to the table that has DEFAULT expression using time-related
builtins like `now()` and `localtimestamp`. The bug was introduced in
23.1.0 and is now fixed.

yuzefovich commented:
Apologies for all the pushing, this is RFAL.

cucaroach reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice work!

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105233#-:-NYQPTMS6f9wRpHKWwC6:b-8cucv6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice find! thanks for fixing this

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105233#-:-NYTRBdr7-4Q90QDTBtr:b-nx7r3b)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+

yuzefovich commented:
bors r+

yuzefovich commented:
Known flake (just borsed the fix).

bors r+
